### ISSUE TITLE
Tips and Tricks for training classification convolutional neural networks

### ISSUE THREAD

**Issue by zaccharieramzi** [benchopt/benchmark_resnet_classif#11]
Data augmentation:

- [ ] Random Resized Crop
- [x] Horizontal Flip
- [ ] Random Augment (see [code](https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/autoaugment.py#L672), [paper](https://arxiv.org/abs/1909.13719) ), with e.g. 7/0.5. Basically it is a fixed set of data augmentation functions.
- [ ] Color Jitter
- [ ] PCA lighting
- [ ] Random Erasing
- [ ] Mixup
- [ ] Cutmix

Regularization:

- [ ] Label smoothing (basically making the target of classif not (1, 0, 0 ,0) but (1-e, e/3, e/3, e/3)
- [ ] Repeated Augmentation (multiple instances of a sample in the same batch, with different augmentation)
- [ ] stochastic depth (basically drop some blocks and replace them with identity stochastically). It sounds very complicated to use as part of a solver in our setup. Moreover, the "ResNet strikes back" paper claims that this is only a good idea for very big networks, so it might not be our priority.
- [x] weight decay (according to "Bag of tricks" this might only be useful for non-BN params). I have seen somewhere (but don't remember where exactly) that it should also only be applied to the weights and not the biases. 

Learning rate:

- [ ] Warmup: scale linearly the learning rate from 0 to the initial in the first few (5) epochs.
- [x] Scheduling: step or cosine. The problem with the cosine schedule is that it needs to know the total number of epochs, which in the current setting is not available to the solver. Could we work around that @tomMoral ?

Modeling (to me these ones are out of our scope):

- [ ] Zero gamma: make the initialization of the learned scale of BN layers to 0.
- [ ] Layer Scale (basically learn a multiplicative factor per channel at the end of residual blocks)

Other:

- [ ] mixed precision
- [ ] weight averaging (SWA or EMA)
- [ ] Binary Xent rather than categorical Xent, coupled with Cutmix and Mixup, in a 1-vs-all fashion
- [ ] gradient clipping (not used in "ResNet strikes back")

**Comment by zaccharieramzi** [comment]
RE LR scheduling: there exists a big difference in how TF and PL implement it.
Basically, TF implements it on a per-optimizer-step level, and PL implements it on a per-epoch level (see [here](https://pytorch-lightning.readthedocs.io/en/latest/common/optimization.html#automatic-optimization)), which is similar that is done [here](https://github.com/kuangliu/pytorch-cifar/blob/master/main.py#L154) or in [timm](https://github.com/rwightman/pytorch-image-models/blob/master/train.py#L650).

I might just go with the PL way of doing it, since it's the least flexible.

**Comment by zaccharieramzi** [comment]
RE LR scheduling / Weight Decay: I am not sure what is the canonical way of updating the weight decay given the lr schedule.

In TF, [it is specified](https://www.tensorflow.org/addons/api_docs/python/tfa/optimizers/extend_with_decoupled_weight_decay) that it should be updated, and in this case manually: 
>Note: when applying a decay to the learning rate, be sure to manually apply the decay to the `weight_decay` as well. For example:

But in PL or torch, I didn't see any mention of this update, so it might not be used. I am going to verify this.
